src/transaction/log_recovery_redo_parallel.cpp

```diff
@@ -0,0 +1,386 @@
+/*
+ * Copyright 2008 Search Solution Corporation
+ * Copyright 2016 CUBRID Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#include "vpid.hpp"
+#include "log_recovery_redo_parallel.hpp"
+
+namespace cublog
+{
+  /*********************************************************************
+   * redo_parallel::redo_job_queue
+   *********************************************************************/
+
+  redo_parallel::redo_job_queue::redo_job_queue ()
+    : m_produce_queue ( new ux_redo_job_deque () )
+    , m_consume_queue ( new ux_redo_job_deque () )
+    , m_queues_empty (false)
+    , m_adding_finished { false }
+  {
+  }
+
+  redo_parallel::redo_job_queue::~redo_job_queue ()
+  {
+    assert (m_produce_queue->size () == 0);
+    assert (m_consume_queue->size () == 0);
+    assert (m_in_progress_vpids.size () == 0);
+
+    delete m_produce_queue;
+    m_produce_queue = nullptr;
+
+    delete m_consume_queue;
+    m_consume_queue = nullptr;
+  }
+
+  void redo_parallel::redo_job_queue::locked_push (ux_redo_job_base &&job)
+  {
+    std::lock_guard<std::mutex> lck (m_produce_queue_mutex);
+    m_produce_queue->push_back (std::move (job));
+    m_queues_empty = false;
+  }
+
+  void redo_parallel::redo_job_queue::set_adding_finished ()
+  {
+    m_adding_finished.store (true);
+  }
+
+  bool redo_parallel::redo_job_queue::get_adding_finished () const
+  {
+    return m_adding_finished.load ();
+  }
+
+  redo_parallel::redo_job_queue::ux_redo_job_base redo_parallel::redo_job_queue::locked_pop (bool &out_adding_finished)
+  {
+    std::unique_lock<std::mutex> consume_queue_lock (m_consume_queue_mutex);
+
+    out_adding_finished = false;
```

Comment:
May be loaded here directly.
```suggestion
    out_adding_finished = m_adding_finished.load ();
```

Reply:
I did it like this, conservatively, on purpose. The out param is set only when it is needed to be checked. There's a sort of hierarchy between the return value and the out param value:
- the return value has higher prio
- the out param is to be checked only when the return value is null
Ideally, it should be refactored into a semantically stronger return type.
I can do it if needed.

Reply:
Later edit.
I've added a proper explanation why it is loaded only in one situation:
```
	      // specifically leave the 'out_adding_finished' on false:
	      //  - even if adding has been finished, the produce queue might still have jobs to be consumed
	      //  - if so, setting the out param to true here, will cause consuming tasks to finish
	      //  - thus allowing for a corner case where the jobs are not fully processed and all
	      //    tasks have finished executing
```